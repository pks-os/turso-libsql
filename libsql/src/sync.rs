const DEFAULT_MAX_RETRIES: usize = 5;
pub struct SyncContext {
    pub sync_url: String,
    pub auth_token: Option<String>,
    pub max_retries: usize,
    pub durable_frame_num: u32,
}

impl SyncContext {
    pub fn new(sync_url: String, auth_token: Option<String>) -> Self {
        Self {
            sync_url,
            auth_token,
            durable_frame_num: 0,
            max_retries: DEFAULT_MAX_RETRIES,
        }
    }
}
