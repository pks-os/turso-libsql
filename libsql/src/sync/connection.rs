use crate::{
    connection::Conn,
    hrana::{connection::HttpConnection, hyper::HttpSender},
    local::{self, impls::LibsqlStmt},
    params::Params,
    replication::connection::State,
    sync::SyncContext,
    BatchRows, Error, Result, Statement, Transaction, TransactionBehavior,
};
use std::sync::Arc;
use tokio::sync::Mutex;

use super::{statement::SyncedStatement, transaction::SyncedTx};

#[derive(Clone)]
pub struct SyncedConnection {
    pub remote: HttpConnection<HttpSender>,
    pub local: local::Connection,
    pub read_your_writes: bool,
    pub context: Arc<Mutex<SyncContext>>,
    pub state: Arc<Mutex<State>>,
}

impl SyncedConnection {
    async fn should_execute_local(&self, sql: &str) -> Result<bool> {
        let stmts = crate::parser::Statement::parse(sql)
            .collect::<Result<Vec<_>>>()
            .or_else(|err| match err {
                Error::Sqlite3UnsupportedStatement => Ok(vec![]),
                err => Err(err),
            })?;

        let mut state = self.state.lock().await;

        crate::replication::connection::should_execute_local(&mut state, stmts.as_slice())
    }
}

#[async_trait::async_trait]
impl Conn for SyncedConnection {
    async fn execute(&self, sql: &str, params: Params) -> Result<u64> {
        let mut stmt = self.prepare(sql).await?;
        stmt.execute(params).await.map(|v| v as u64)
    }

    async fn execute_batch(&self, sql: &str) -> Result<BatchRows> {
        if self.should_execute_local(sql).await? {
            self.local.execute_batch(sql)
        } else {
            self.remote.execute_batch(sql).await
        }
    }

    async fn execute_transactional_batch(&self, sql: &str) -> Result<BatchRows> {
        if self.should_execute_local(sql).await? {
            self.local.execute_transactional_batch(sql)?;
            Ok(BatchRows::empty())
        } else {
            self.remote.execute_transactional_batch(sql).await
        }
    }

    async fn prepare(&self, sql: &str) -> Result<Statement> {
        if self.should_execute_local(sql).await? {
            Ok(Statement {
                inner: Box::new(LibsqlStmt(self.local.prepare(sql)?)),
            })
        } else {
            let stmt = Statement {
                inner: Box::new(self.remote.prepare(sql)?),
            };

            if self.read_your_writes {
                Ok(Statement {
                    inner: Box::new(SyncedStatement {
                        conn: self.local.clone(),
                        context: self.context.clone(),
                        inner: stmt,
                    }),
                })
            } else {
                Ok(stmt)
            }
        }
    }

    async fn transaction(&self, tx_behavior: TransactionBehavior) -> Result<Transaction> {
        let tx = SyncedTx::begin(self.clone(), tx_behavior).await?;

        Ok(Transaction {
            inner: Box::new(tx),
            conn: crate::Connection {
                conn: Arc::new(self.clone()),
            },
            close: None,
        })
    }

    fn interrupt(&self) -> Result<()> {
        Ok(())
    }

    fn is_autocommit(&self) -> bool {
        self.remote.is_autocommit()
    }

    fn changes(&self) -> u64 {
        self.remote.changes()
    }

    fn total_changes(&self) -> u64 {
        self.remote.total_changes()
    }

    fn last_insert_rowid(&self) -> i64 {
        self.remote.last_insert_rowid()
    }

    async fn reset(&self) {}
}
