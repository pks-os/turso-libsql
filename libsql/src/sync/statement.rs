use crate::{
    local::{self},
    params::Params,
    statement::Stmt,
    sync::SyncContext, Column, Result, Rows, Statement,
};
use std::sync::Arc;
use tokio::sync::Mutex;

pub struct SyncedStatement {
    pub conn: local::Connection,
    pub context: Arc<Mutex<SyncContext>>,
    pub inner: Statement,
}

#[async_trait::async_trait]
impl Stmt for SyncedStatement {
    fn finalize(&mut self) {
        self.inner.finalize()
    }

    async fn execute(&mut self, params: &Params) -> Result<usize> {
        let result = self.inner.execute(params).await;
        let mut context = self.context.lock().await;
        let _ = crate::sync::sync_offline(&mut context, &self.conn).await;
        result
    }

    async fn query(&mut self, params: &Params) -> Result<Rows> {
        let result = self.inner.query(params).await;
        let mut context = self.context.lock().await;
        let _ = crate::sync::sync_offline(&mut context, &self.conn).await;
        result
    }

    async fn run(&mut self, params: &Params) -> Result<()> {
        let result = self.inner.run(params).await;
        let mut context = self.context.lock().await;
        let _ = crate::sync::sync_offline(&mut context, &self.conn).await;
        result
    }

    fn reset(&mut self) {
        self.inner.reset()
    }

    fn parameter_count(&self) -> usize {
        self.inner.parameter_count()
    }

    fn parameter_name(&self, idx: i32) -> Option<&str> {
        self.inner.parameter_name(idx)
    }

    fn columns(&self) -> Vec<Column> {
        self.inner.columns()
    }
}
