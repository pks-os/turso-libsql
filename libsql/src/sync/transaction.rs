use crate::{
    connection::Conn,
    params::Params,
    transaction::Tx, Result, TransactionBehavior,
};

use super::connection::SyncedConnection;

pub struct SyncedTx(SyncedConnection);

impl SyncedTx {
    pub(crate) async fn begin(
        conn: SyncedConnection,
        tx_behavior: TransactionBehavior,
    ) -> Result<Self> {
        conn.execute(
            match tx_behavior {
                TransactionBehavior::Deferred => "BEGIN DEFERRED",
                TransactionBehavior::Immediate => "BEGIN IMMEDIATE",
                TransactionBehavior::Exclusive => "BEGIN EXCLUSIVE",
                TransactionBehavior::ReadOnly => "BEGIN READONLY",
            },
            Params::None,
        )
        .await?;
        Ok(Self(conn.clone()))
    }
}

#[async_trait::async_trait]
impl Tx for SyncedTx {
    async fn commit(&mut self) -> Result<()> {
        self.0.execute("COMMIT", Params::None).await?;
        Ok(())
    }

    async fn rollback(&mut self) -> Result<()> {
        self.0.execute("ROLLBACK", Params::None).await?;
        Ok(())
    }
}
