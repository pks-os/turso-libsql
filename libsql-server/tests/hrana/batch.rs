use std::sync::Arc;
use std::time::Duration;

use futures::StreamExt;
use insta::{assert_json_snapshot, assert_snapshot};
use libsql::{params, Database};
use libsql_server::config::UserApiConfig;
use libsql_server::hrana_proto::{Batch, BatchStep, Stmt};
use tokio::sync::Notify;

use crate::common::http::Client;
use crate::common::net::{init_tracing, SimServer as _, TestServer, TurmoilConnector};

#[test]
fn sample_request() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let batch = Batch {
            steps: vec![BatchStep {
                condition: None,
                stmt: Stmt {
                    sql: Some("create table test (x)".to_string()),
                    ..Default::default()
                },
            }],
            replication_index: None,
        };
        let client = Client::new();

        let resp = client
            .post(
                "http://primary:8080/v1/batch",
                serde_json::json!({ "batch": batch }),
            )
            .await
            .unwrap();

        let mut json = resp.json_value().await.unwrap();

        for result in json["result"]["step_results"]
            .as_array_mut()
            .unwrap()
            .iter_mut()
        {
            result
                .as_object_mut()
                .unwrap()
                .remove("query_duration_ms")
                .expect("expected query_duration_ms");
        }

        assert_json_snapshot!(json);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn execute_individual_statements() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute("create table t(x text)", ()).await?;
        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;
        let mut rows = conn
            .query("select * from t where x = ?", params!["hello"])
            .await?;

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.column_name(0), Some("x"));
        assert_eq!(rows.next().await?.unwrap().get::<String>(0)?, "hello");
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn execute_batch() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute_batch(
            r#"
        begin;
        create table t(x text);
        insert into t(x) values('hello; world');
        end;"#,
        )
        .await?;
        let mut rows = conn
            .query("select * from t where x = ?", params!["hello; world"])
            .await?;

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.column_name(0), Some("x"));
        assert_eq!(
            rows.next().await?.unwrap().get::<String>(0)?,
            "hello; world"
        );
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn execute_batch_returning() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        let mut batch_rows = conn
            .execute_transactional_batch(
                r#"
        create table t(x text);
        insert into t(x) values('hello; world') RETURNING *;
        "#,
            )
            .await?;

        batch_rows.next_stmt_row();

        let mut rows = batch_rows.next_stmt_row().unwrap().unwrap();

        assert!(batch_rows.next_stmt_row().is_none());

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.column_name(0), Some("x"));
        assert_eq!(
            rows.next().await?.unwrap().get::<String>(0)?,
            "hello; world"
        );
        assert!(rows.next().await?.is_none());

        let mut batch_rows = conn
            .execute_batch(
                r#"
        create table t2(x text);
        insert into t2(x) values('hello; world') RETURNING *;
        "#,
            )
            .await?;

        batch_rows.next_stmt_row();

        let mut rows = batch_rows.next_stmt_row().unwrap().unwrap();

        assert!(batch_rows.next_stmt_row().is_none());

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.column_name(0), Some("x"));
        assert_eq!(
            rows.next().await?.unwrap().get::<String>(0)?,
            "hello; world"
        );
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn multistatement_query() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;
        let mut rows = conn
            .query("select 1 + ?; select 'abc';", params![1])
            .await?;

        assert_eq!(rows.column_count(), 1);
        assert_eq!(rows.next().await?.unwrap().get::<i32>(0)?, 2);
        assert!(rows.next().await?.is_none());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn affected_rows_and_last_rowid() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute(
            "create table t(id integer primary key autoincrement, x text);",
            (),
        )
        .await?;

        let r = conn.execute("insert into t(x) values('a');", ()).await?;
        assert_eq!(r, 1, "1st row inserted");
        assert_eq!(conn.last_insert_rowid(), 1, "1st row id");

        let r = conn
            .execute("insert into t(x) values('b'),('c');", ())
            .await?;
        assert_eq!(r, 2, "2nd and 3rd rows inserted");
        assert_eq!(conn.last_insert_rowid(), 3, "3rd row id");

        let r = conn.execute("update t set x = 'd';", ()).await?;
        assert_eq!(r, 3, "all three rows updated");
        assert_eq!(conn.last_insert_rowid(), 3, "last row id unchanged");

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn stats() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let req = serde_json::json!({
            "requests": [
                {"type": "execute", "stmt": { "sql": "CREATE TABLE foo (x INT)" }},
                {"type": "execute", "stmt": { "sql": "INSERT INTO foo VALUES (42)"}},
                {"type": "execute", "stmt": { "sql": "SELECT * FROM foo"}}
            ]
        });
        let client = Client::new();

        let resp = client
            .post("http://primary:8080/v2/pipeline", req)
            .await
            .unwrap();

        let mut json = resp.json_value().await.unwrap();
        json.as_object_mut().unwrap().remove("baton");

        for results in json["results"].as_array_mut().unwrap().iter_mut() {
            results["response"]["result"]
                .as_object_mut()
                .unwrap()
                .remove("query_duration_ms")
                .expect("expected query_duration_ms");
        }

        assert_json_snapshot!(json);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn stats_legacy() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let req = serde_json::json!({
            "statements": [
                "CREATE TABLE foo (x INT)",
                "INSERT INTO foo VALUES (42)",
                "SELECT * FROM foo"
            ]
        });
        let client = Client::new();

        let resp = client.post("http://primary:8080/", req).await.unwrap();

        let mut json = resp.json_value().await.unwrap();

        for result in json.as_array_mut().unwrap() {
            result["results"]
                .as_object_mut()
                .unwrap()
                .remove("query_duration_ms")
                .expect("expected query_duration_ms");
        }

        assert_json_snapshot!(json);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn stream() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute("create table t(x text)", ()).await?;
        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;

        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;

        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;

        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;

        let rows = conn
            .query("select * from t where x = ?", params!["hello"])
            .await?
            .into_stream();

        let rows = rows.collect::<Vec<_>>().await;

        assert_eq!(rows.len(), 4);

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn reindex_statement() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute("create table t(x text)", ()).await?;
        conn.execute("create index t_idx on t(x)", ()).await?;
        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;
        conn.execute("insert into t(x) values(?)", params!["hello"])
            .await?;
        conn.execute("reindex t_idx", ()).await?;

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn test_simulate_vector_index_load_from_dump() {
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();
    sim.host("primary", super::make_standalone_server);
    sim.client("client", async {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;
        let conn = db.connect()?;

        conn.execute("CREATE TABLE t ( v FLOAT32(2) );", ()).await?;
        conn.execute("CREATE TABLE t_idx_shadow(index_key  INTEGER , data BLOB, PRIMARY KEY (index_key));", ()).await?;
        conn.execute("CREATE TABLE libsql_vector_meta_shadow ( name TEXT PRIMARY KEY, metadata BLOB ) WITHOUT ROWID", ()).await?;
        conn.execute("INSERT INTO libsql_vector_meta_shadow VALUES ('t_idx', x'');", ()).await?;
        conn.execute("INSERT INTO t VALUES (vector('[1,2]')), (vector('[2,3]'));", ()).await?;
        conn.execute("CREATE INDEX t_idx ON t (libsql_vector_idx(v));", ()).await?;

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
fn server_restart_query_execute_invalid_baton() {
    server_restart(|notify, db| async move {
        let conn = db.connect().unwrap();

        conn.query("select 1;", ()).await.unwrap();

        notify.notify_waiters();
        tokio::time::sleep(std::time::Duration::from_secs(1)).await;

        let err = conn.execute("select 1;", ()).await.unwrap_err();
        assert_snapshot!(err);

        Ok(())
    });
}

#[test]
fn server_restart_txn_execute_execute_invalid_baton() {
    server_restart(|notify, db| async move {
        let conn = db.connect().unwrap();

        let txn = conn.transaction().await.unwrap();

        txn.execute("select 1;", ()).await.unwrap();

        notify.notify_waiters();

        tokio::time::sleep(std::time::Duration::from_secs(1)).await;

        let err = txn.execute("select 1;", ()).await.unwrap_err();

        assert_snapshot!(err);

        Ok(())
    });
}

#[test]
fn server_restart_txn_query_execute_invalid_baton() {
    server_restart(|notify, db| async move {
        let conn = db.connect().unwrap();

        let txn = conn.transaction().await.unwrap();

        txn.query("select 1;", ()).await.unwrap();

        notify.notify_waiters();

        tokio::time::sleep(std::time::Duration::from_secs(1)).await;

        let err = txn.execute("select 1;", ()).await.unwrap_err();

        assert_snapshot!(err);

        Ok(())
    });
}

#[test]
fn server_restart_txn_execute_query_invalid_baton() {
    server_restart(|notify, db| async move {
        let conn = db.connect().unwrap();

        let txn = conn.transaction().await.unwrap();

        txn.execute("select 1;", ()).await.unwrap();

        notify.notify_waiters();

        tokio::time::sleep(std::time::Duration::from_secs(1)).await;

        let err = txn.query("select 1;", ()).await.unwrap_err();

        assert_snapshot!(err);

        Ok(())
    });
}

#[test]
fn server_restart_execute_query() {
    server_restart(|notify, db| async move {
        let conn = db.connect().unwrap();
        conn.execute("select 1;", ()).await.unwrap();

        notify.notify_waiters();

        tokio::time::sleep(std::time::Duration::from_secs(1)).await;

        conn.query("select 1;", ()).await.unwrap();

        Ok(())
    });
}

#[test]
fn server_timeout() {
    server_restart(|_notify, db| async move {
        let conn = db.connect().unwrap();
        conn.query("select 1;", ()).await.unwrap();

        tokio::time::sleep(std::time::Duration::from_secs(20)).await;

        let err = conn.execute("select 1;", ()).await.unwrap_err();

        assert_snapshot!(err);

        let conn = db.connect().unwrap();
        conn.execute("select 1;", ()).await.unwrap();

        tokio::time::sleep(std::time::Duration::from_secs(20)).await;

        conn.execute("select 1;", ()).await.unwrap();

        Ok(())
    });
}

#[track_caller]
fn server_restart<F, Fut>(f: F)
where
    F: Fn(Arc<Notify>, Database) -> Fut + 'static,
    Fut: std::future::Future<Output = Result<(), libsql::Error>> + 'static,
{
    let mut sim = turmoil::Builder::new()
        .simulation_duration(Duration::from_secs(1000))
        .build();

    init_tracing();

    let notify = Arc::new(Notify::new());

    let notify_clone = notify.clone();

    sim.host("primary", move || {
        let notify = notify.clone();
        async move {
            let tmp = tempfile::tempdir()?;

            let make_server = || TestServer {
                path: tmp.path().to_owned().into(),
                user_api_config: UserApiConfig {
                    hrana_ws_acceptor: None,
                    ..Default::default()
                },
                ..Default::default()
            };

            let server = make_server();

            tokio::select! {
                res = server.start_sim(8080) => {
                    res.unwrap()
                }
                _ = notify.notified() => (),
            }

            let server = make_server();
            server.start_sim(8080).await.unwrap();

            Ok(())
        }
    });

    sim.client("client", async move {
        let db = Database::open_remote_with_connector("http://primary:8080", "", TurmoilConnector)?;

        f(notify_clone, db).await.unwrap();

        Ok(())
    });

    sim.run().unwrap();
}
