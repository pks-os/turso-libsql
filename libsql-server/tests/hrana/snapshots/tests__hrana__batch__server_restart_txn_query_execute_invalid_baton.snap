---
source: libsql-server/tests/hrana/batch.rs
expression: err
snapshot_kind: text
---
Hrana: `api error: `status=400 Bad Request, body=Received an invalid baton``
